lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -29,4 +132,273 @@
+    """This class is used to manage the routing table of the router provider.
+
+    It's capabilities are to:
+    * Manage the routing table in the charm itself, by adding and removing
+    new network and route requests by integrated units,
+    * Synchronize the databags of all requiring units with the router table of the
+    provider charm
+
+    Attributes:
+        charm:
+            The Charm object that instantiates this class
+        _stored:
+            The persistent state that keeps the internal routing table.
+    """
+
+    on = RouterProviderCharmEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "ip-router")
+        self.charm = charm
+        self._stored.set_default(routing_table={})
+        self.framework.observe(
+            charm.on.ip_router_relation_changed, self._on_ip_router_relation_changed
+        )
+        self.framework.observe(
+            charm.on.ip_router_relation_joined, self._on_ip_router_relation_joined
+        )
+        self.framework.observe(
+            charm.on.ip_router_relation_departed, self._on_ip_router_relation_departed
+        )
+
+    def _on_ip_router_relation_joined(self, event: RelationJoinedEvent):
+        """When a new unit or app joins the relation, add its name to the routing table"""
+        self._stored.routing_table.update({event.relation.app.name: []})
+        self._sync_routing_tables()
+
+    def _on_ip_router_relation_changed(self, event: RelationChangedEvent):
+        """Update the internal routing table state to reflect changes in
+        requirer units' new network requests.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if "networks" not in event.relation.data[event.relation.app]:
+            return
+
+        new_network = event.relation.data[event.relation.app]["networks"]
+        if new_network == self._stored.routing_table[event.relation.app.name]:
+            return
+        self._stored.routing_table[event.relation.app.name] = new_network
+        self._sync_routing_tables()
+        self.on.routing_table_updated.emit()
+
+    def _on_ip_router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if len(event.relation.units) == 0:
+            self._stored.routing_table.pop(event.app.name)
+            self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Read-only way to get the current routing table"""
+        return deepcopy(self._stored.routing_table._under)  # RFC: Is there a better way?
+
+    def get_flattened_routing_table(self) -> List[Network]:
+        """Returns a read-only internal routing table that's flattened
+
+        Returns:
+            A list of objects of type `Network`
+        """
+        internal_rt = self.get_routing_table()
```

Comment:
Avoid acronyms like "rt", "internal_routing_table" is fine.

Reply:
Done